### PR TITLE
Feature: Adding Task Manager

### PR DESCRIPTION
**Pull request by ahmadnazeri**
# Description

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have fixed any merge conflicts

### CODE REVIEW

File: scheduling/storage_provider.go [hunk 1]
```diff
@@ -2,12 +2,14 @@ package scheduling
 
 import (
 	"fmt"
+	"sort"
 	"strings"
 )
 
 type StorageProvider interface {
 	Set(key string, value string) error
 	Get(key string, prefix bool) ([]string, error)
```
Comment (aolfat) [inline scheduling/storage_provider.go:11]: what's the bool for prefix here?

File: scheduling/storage_provider.go [hunk 2]
```diff
@@ -2,12 +2,14 @@ package scheduling
 
 import (
 	"fmt"
+	"sort"
 	"strings"
 )
 
 type StorageProvider interface {
```
Comment (aolfat) [inline scheduling/storage_provider.go:9]: does it make sense to store StorageProvider in scheduling? Won't it eventually be the abstraction used in coordinator as well?

File: scheduling/task_run_list.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+)
+
+type TaskRunSimple struct {
+	RunID       TaskRunID `json:"runID"`
+	DateCreated time.Time `json:"dateCreated"`
```
Comment (aolfat) [inline scheduling/task_run_list.go:11]: nit: CreatedAt

File: scheduling/storage_provider.go [hunk 4]
```diff
@@ -41,9 +43,18 @@ func (m *MemoryStorageProvider) Get(key string, prefix bool) ([]string, error) {
...
 		if strings.HasPrefix(k, key) {
-			result = append(result, v)
+			keys = append(keys, k)
+		}
+	}
+
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		if strings.HasPrefix(k, key) {
```
Comment (anthonylasso) [inline scheduling/storage_provider.go:56]: I may be missing context, but do we need this second `.hasPrefix` check if keys is already made of items with the prefix?

File: scheduling/task_run_list.go (hunk 3 above)
Comment (anthonylasso) [reply scheduling/task_run_list.go:11]: we're already using `dateCreated` throughout

File: scheduling/taskmanager.go [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+}
+
+type TaskManager struct {
+	storage StorageProvider
+}
+
+type TaskMetadataList []TaskMetadata
+
+func (tml *TaskMetadataList) ToJSON() string {
+	return ""
```
Comment (anthonylasso) [inline scheduling/taskmanager.go:69]: todo?

File: scheduling/taskmanager.go [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+	err = tm.storage.Set(taskRunKey.String(), string(serializedRuns))
+	if err != nil {
+		return TaskMetadata{}, err
+	}
+
+	return metadata, nil
+}
+
+// Finds the highest increment in a list of strings formatted like "/tasks/metadata/task_id=0"
+func getLatestID(taskPaths []string) (int, error) {
```
Comment (anthonylasso) [inline scheduling/taskmanager.go:129]: will accidental dupes be an issue? wondering if concurrent calls might result in the same IDs getting returned.

File: scheduling/taskmanager.go [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+		err = taskMetadata.Unmarshal([]byte(m))
+		if err != nil {
+			return TaskMetadataList{}, err
+		}
+		tml = append(tml, taskMetadata)
+	}
+	return tml, nil
+}
+
+// Need to implement this
```
Comment (anthonylasso) [inline scheduling/taskmanager.go:189]: still need this?

File: scheduling/storage_providers/memory.go
Comment (sdreyer) [inline scheduling/storage_providers/memory.go:1]: Can we move memory storage into the memory file and just have the interface in this one?

File: scheduling/storage_providers/memory.go [hunk 8]
```diff
@@ -2,18 +2,51 @@ package scheduling
...
 
 type MemoryStorageProvider struct {
-	storage map[string]string
+	storage     map[string]string
+	lockedItems map[string]LockInformation
 }
 
-func NewMemoryStorageProvider() *MemoryStorageProvider {
-	return &MemoryStorageProvider{storage: make(map[string]string)}
+func NewMemoryStorageProvider(channel chan int) *MemoryStorageProvider {
```
Comment (sdreyer) [inline scheduling/storage_providers/memory.go:27]: Do we need this? can we just create this inside and return it?

Comment (sdreyer) [reply scheduling/storage_providers/memory.go:27]: Referring to the channel

File: scheduling/storage_providers/memory.go [hunk 9]
```diff
@@ -37,13 +70,109 @@ func (m *MemoryStorageProvider) Get(key string, prefix bool) ([]string, error) {
...
+
+	return result, nil
+}
+
+type LockObject struct {
+	ID      string
+	Channel chan error
+}
+
+func (m *MemoryStorageProvider) Lock(id string, key string, lockChannel chan error) (LockObject, error) {
```
Comment (sdreyer) [inline scheduling/storage_providers/memory.go:110]: can we create the lock internally and return it? Dont want to have to pass it in

File: scheduling/storage_providers/memory.go [hunk 10]
```diff
@@ -2,48 +2,186 @@ package scheduling
...
 	"fmt"
+	"sort"
 	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+)
+
+type LockInformation struct {
```
Comment (aolfat) [inline scheduling/storage_providers/memory.go:13]: nit: move LockObject right above this

File: scheduling/storage_providers/memory.go [hunk 11]
```diff
@@ -2,48 +2,186 @@ package scheduling
...
+	ID   string
+	Key  string
+	Date time.Time
+	Lock LockObject
+}
+
+const (
+	UpdateSleepTime = 2 * time.Second
+	ValidTimePeriod = 5 * time.Second
+	CleanupInterval = 10 * time.Second
```
Comment (aolfat) [inline scheduling/storage_providers/memory.go:23]: remove

File: scheduling/taskmanager.go (hunk 5 above)
Comment (anthonylasso) [reply scheduling/taskmanager.go:70]: I don't think I need this, so resolving for now.

File: scheduling/storage_providers/memory.go [hunk 12]
```diff
@@ -2,48 +2,186 @@ package scheduling
...
-func (m *MemoryStorageProvider) Set(key string, value string) error {
+func (m *MemoryStorageProvider) Set(key string, value string, lock LockObject) error {
+	lockInfo, ok := m.lockedItems.Load(key)
+	if !ok {
+		return fmt.Errorf("key is not locked")
+	}
+
+	currentLock := lockInfo.(LockInformation)
+	if currentLock.ID != lock.ID {
+		return fmt.Errorf("key is locked by another id: locked by: %s, unlock by: %s", currentLock.ID, lock.ID)
```
Comment (anthonylasso) [inline scheduling/storage_providers/memory.go:45]: nit: include the key string in the error msg

File: scheduling/storage_providers/memory.go [hunk 13]
```diff
@@ -2,48 +2,186 @@ package scheduling
...
+	return result, nil
+}
+
+type LockObject struct {
+	ID      string
+	Channel *chan error
+}
+
+func (m *MemoryStorageProvider) Lock(key string) (LockObject, error) {
+	id := uuid.New().String()
```
Comment (anthonylasso) [inline scheduling/storage_providers/memory.go:112]: nit: you can save a few machine cycles by doing the id assignment until after the key == "" validation check.

**Review by anthonylasso — APPROVED** [review]
Giant lift Ahmad! 🏋️ Left a few nits, edit as needed.

File: scheduling/storage_providers/etcd_test.go [hunk 14]
```diff
@@ -1,43 +1,38 @@
 package scheduling
```
[the remark this answers is not in the document]
Comment (ahmadnazeri) [reply scheduling/storage_providers/etcd_test.go:1]: I will need to modify it for etcd so didn't want to delete it.

File: scheduling/storage_providers/storage_provider_test.go [hunk 15]
```diff
@@ -87,16 +101,197 @@ func StorageProviderGet(t *testing.T, provider StorageProvider) {
...
+
 	for i := range a {
 		if a[i] != b[i] {
 			return false
 		}
 	}
 	return true
 }
+
+func compareMaps(a, b map[string]string) bool {
```
Comment (aolfat) [inline scheduling/storage_providers/storage_provider_test.go:166]: should you use reflect.DeepEqual here?

File: scheduling/storage_providers/storage_provider_test.go [hunk 16]
```diff
@@ -87,16 +101,197 @@ func StorageProviderGet(t *testing.T, provider StorageProvider) {
...
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			runTestCase(t, test)
+		})
+	}
+}
+
 func compareStringSlices(a, b []string) bool {
```
Comment (aolfat) [inline scheduling/storage_providers/storage_provider_test.go:151]: some thing here, should this just be reflect.DeepEqual

File: scheduling/storage_providers/memory.go [hunk 17]
```diff
@@ -2,48 +2,187 @@ package scheduling
...
+		Date: time.Now(),
+		Lock: lockObject,
+	}
+	m.lockedItems.Store(key, lock)
+
+	return lockObject, nil
+}
+
+func (m *MemoryStorageProvider) Unlock(key string, lock LockObject) error {
+	lockMutex := &sync.Mutex{}
```
Comment (aolfat) [inline scheduling/storage_providers/memory.go:145]: wait. I don't think this is right.

What does the lockMutex do here?

You create a new lockMutex on every call to this function. So... each call to this function has a new mutex? Doesn't there need to be shared mutex for this all to work?

File: scheduling/storage_providers/memory.go [hunk 18]
```diff
@@ -2,48 +2,187 @@ package scheduling
...
+	defer lockMutex.Unlock()
+
+	if lockInfo, ok := m.lockedItems.Load(key); ok {
+		keyLock := lockInfo.(LockInformation)
+		if time.Since(keyLock.Date) < ValidTimePeriod {
+			return LockObject{}, fmt.Errorf("key is already locked by: %s", keyLock.ID)
+		}
+	}
+
+	lockChannel := make(chan error)
```
Comment (aolfat) [inline scheduling/storage_providers/memory.go:129]: we never close the lockchannel

**Pull request merged**